### PR TITLE
[Feature]: Support pre vote phase

### PR DESCRIPTION
**Pull request by themanforfree**
Please briefly answer these questions:
Support pre vote phase
* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Avoid timed out nodes from disrupting a healthy cluster
* what changes does this pull request make?
Add pre vote and add more check when node receive a vote request
* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)
no

### CODE REVIEW

File: curp/src/server/raw_curp/mod.rs [hunk 1]
```diff
@@ -519,6 +530,51 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
...
+    /// Return `Ok(term, spec_pool)` if the vote is granted
+    /// Return `Err(term)` if the vote is rejected
+    pub(super) fn handle_pre_vote(
+        &self,
+        term: u64,
+        candidate_id: ServerId,
+        last_log_index: LogIndex,
+        last_log_term: u64,
+    ) -> Result<(u64, Vec<PoolEntry<C>>), u64> {
+        debug!(
```
Comment (bsbds) [inline curp/src/server/raw_curp/mod.rs:543]: Please also update this debug message.

File: curp/src/server/raw_curp/mod.rs [hunk 2]
```diff
@@ -519,6 +530,51 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
...
+        }
+
+        // check if the candidate's log is up-to-date
+        if !log_r.log_up_to_date(last_log_term, last_log_index) {
+            return Err(st_r.term);
+        }
+
+        // grant the vote
+        debug!("{} pre votes for server {}", self.id(), candidate_id);
+        self.reset_election_tick();
```
Comment (bsbds) [inline curp/src/server/raw_curp/mod.rs:574]: If we reset election tick here, will the follower accept the pre-vote but reject the vote?

File: curp/src/server/raw_curp/mod.rs [hunk 3]
```diff
@@ -519,6 +530,51 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
         Ok((st_w.term, self_spec_pool))
     }
 
+    /// Handle `pre_vote`
+    /// Return `Ok(term, spec_pool)` if the vote is granted
+    /// Return `Err(term)` if the vote is rejected
+    pub(super) fn handle_pre_vote(
```
Comment (iGxnon) [inline curp/src/server/raw_curp/mod.rs:536]: Could we merge this method into `handle_vote()`? They are very similar.

File: curp/src/server/raw_curp/mod.rs [hunk 4]
```diff
@@ -489,6 +495,11 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
             return Err(st_w.term);
         }
         if term > st_w.term {
+            let timeout = st_w.follower_timeout_ticks;
+            if st_w.leader_id.is_some() && self.ctx.election_tick.load(Ordering::Acquire) < timeout
```
Comment (iGxnon) [inline curp/src/server/raw_curp/mod.rs:499]: We already check that the leader is None in `PreVote` phase, as the entire process is `PreVote` -> `Vote`, should we check it again here?

File: curp/src/server/raw_curp/mod.rs (hunk 1 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:543]: Updated

File: curp/src/server/raw_curp/mod.rs (hunk 2 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs]: There really shouldn't be a reset here. Removed

File: curp/src/server/raw_curp/mod.rs (hunk 3 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs:536]: I don't think so. They are somewhat similar, but there are more differences, and it is easier to understand in two methods

File: curp/src/server/raw_curp/mod.rs (hunk 4 above)
Comment (themanforfree) [reply curp/src/server/raw_curp/mod.rs]: Removed

**Review by iGxnon — APPROVED** [review]

**Review by bsbds — APPROVED** [review]

**Pull request merged**